presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java

```diff
@@ -61,19 +59,17 @@ public PartitioningExchanger(
             hashGenerator = new InterpretedHashGenerator(partitionChannelTypes, Ints.toArray(partitionChannels));
         }
         partitionGenerator = new LocalPartitionGenerator(hashGenerator, buffers.size());
-
-        partitionAssignments = new IntList[partitions.size()];
-        for (int i = 0; i < partitionAssignments.length; i++) {
-            partitionAssignments[i] = new IntArrayList();
-        }
     }
 
     @Override
-    public synchronized void accept(Page page)
+    public void accept(Page page)
     {
-        // reset the assignment lists
-        for (IntList partitionAssignment : partitionAssignments) {
-            partitionAssignment.clear();
+        int partitionCount = buffers.size();
+
+        IntArrayList[] partitionAssignments = new IntArrayList[partitionCount];
+        int expectedSize = page.getPositionCount() / partitionCount;
```

Comment:
Is it correct to assume even distribution here?
It at least specifies some capacity instead of relying on `IntArrayList` defaults.